bin/ch/WScriptJsrt.cpp

```diff
@@ -1073,6 +1074,115 @@ JsValueRef __stdcall WScriptJsrt::LoadTextFileCallback(JsValueRef callee, bool i
     return returnValue;
 }
 
+int js_fgets(char* buf, int size, FILE* file)
+{
+	int  i, n, c;
+	bool crflag;
+
+	n = size - 1;
+	if (n < 0)
+		return -1;
+
+	crflag = false;
+
+	for (i = 0; i < n && (c = getc(file)) != EOF; i++) {
+		buf[i] = (char)c;
+		if (c == '\n') {        // any \n ends a line
+			i++;                // keep the \n; we know there is room for \0
+			break;
+		}
+		if (crflag) {           // \r not followed by \n ends line at the \r
+			ungetc(c, file);
+			break;              // and overwrite c in buf with \0
+		}
+		crflag = (c == '\r');
+	}
+
+	buf[i] = '\0';
+	return i;
+}
+
+JsValueRef __stdcall WScriptJsrt::ReadLineStdinCallback(JsValueRef callee, bool isConstructCall, JsValueRef *arguments, unsigned short argumentCount, void *callbackState)
+{
+	HRESULT hr = E_FAIL;
+	JsValueRef returnValue = JS_INVALID_REFERENCE;
+	JsErrorCode errorCode = JsNoError;
+
+#define BUFSIZE 256
+	FILE* from = stdin;
+	int buflength = 0;
+	int bufsize = BUFSIZE;
+	char* buf = (char*)malloc(bufsize);
+	bool sawNewline = false;
+	char* tmp;
+	if (!buf)
+		goto Error;
+
+	int gotlength;
+	while ((gotlength = js_fgets(buf + buflength, bufsize - buflength, from)) > 0) {
+		buflength += gotlength;
+
+		/* Are we done? */
+		if (buf[buflength - 1] == '\n') {
+			buf[buflength - 1] = '\0';
+			sawNewline = true;

```

Comment:
Why not do the buflength -1 here? Unless I'm reading it incorrectly, that's effectively the purpose of this flag.

Reply:
I guess i don't understand the question here. Can you maybe show me what you are suggesting?

Reply:
As per how I understand the function, the only use of sawNewline is to keep track that we have an extra null character at the end of the string, and that buflength represents the string length -1. This means that a line with just "\n" won't go through the buflength == 0 path, despite being 0x0000 0x0000 in the binary representation, with a wstrlen of 0.